src/Sema.zig

```diff
@@ -39116,6 +38600,810 @@ pub fn resolveDeclaredEnum(
+            }
+            break :inst src_len;
+        }
+        break :inst .none;
+    };
+    // ... And attempt to resolve the result.
+    const dest_len_val: Value = val: {
+        if (dest_len != .none) {
+            sa.dest_len = .variable;
+            if (try sema.resolveDefinedValue(block, src, dest_len)) |val| {
+                sa.dest_len = .known;
+                break :val val;
+            }
+        }
+        break :val Value.undef;
+    };
+
+    // Extract any sentinel value from input pointer type information:
+    const src_sent_val: Value = val: {
+        const src_sent_ip: InternPool.Index =
+            if (src_ptr_ty_size == .One and src_ptr_child_ty_tag == .array)
+            sema.pt.zcu.intern_pool.indexToKey(Type.toIntern(src_ptr_child_ty)).array_type.sentinel
+        else
+            sema.pt.zcu.intern_pool.indexToKey(Type.toIntern(src_ptr_ty)).ptr_type.sentinel;
+        if (src_ptr_explicit_len and src_sent_ip != .none) {
+            sa.src_sent = .known;
+            break :val Value.fromInterned(src_sent_ip);
+        }
+        break :val Value.undef;
+    };
+    // Attempt to resolve the value of the destination sentinel:
+    const dest_sent_val: Value = val: {
+        if (dest_sent_opt != .none) {
+            sa.dest_sent = .known;
+            try checkSentinelType(sema, block, dest_sent_src, elem_ty);
+            break :val try sema.resolveConstDefinedValue(block, dest_sent_src, try sema.coerce(block, elem_ty, dest_sent_opt, dest_sent_src), .{
+                .needed_comptime_reason = "slice sentinel must be comptime-known",
+            });
+        }
+        break :val src_sent_val;
+    };
+
+    // Compile error for syntax `@as([]u8, &.{})[0.. :0]`.
+    if (sa.destPtrGainsSentinel() and
+        !dest_ptr_explicit_len and src_ptr_explicit_len)
+    {
+        return sema.fail(block, dest_sent_src, "slice sentinel index always out of bounds", .{});
+    }
+
+    if (try elem_ty.comptimeOnlySema(sema.pt)) require_comptime: {
+        assert(sa.src_ptr == .known);
+        const src_note = src_note: {
+            if (sa.dest_start == .variable) {
+                break :src_note .{ dest_start_src, "start index must be comptime-known" };
+            }
+            if (dest_len_opt != .none and sa.dest_len == .variable) {
+                break :src_note .{ dest_end_src, "length must be comptime-known" };
+            }
+            if (dest_end_opt != .none and sa.dest_end == .variable) {
+                break :src_note .{ dest_end_src, "end index must be comptime-known" };
+            }
+            break :require_comptime;
+        };
+        const msg: *Zcu.ErrorMsg = try sema.errMsg(src, "unable to resolve operand slicing comptime-only type '{}'", .{operand_ty.fmt(sema.pt)});
+        {
+            errdefer msg.destroy(sema.gpa);
+            try sema.errNote(src_note[0], msg, "{s}", .{src_note[1]});
+        }
+        return sema.failWithOwnedErrorMsg(block, msg);
+    }
```

Comment:
This validation is not present in master, and because of this it is quite easy to crash the compiler by slicing `comptime`-only types with runtime-known operands.

For example, compile the example program with `zig build-obj runtime_indices_slicing_comptime_only_pointers.zig`
`runtime_indices_slicing_comptime_only_pointers.zig`:
```zig
export fn end() void {
    comptime var buf: [1]comptime_int = .{0};
    var dest_end: usize = 0;
    dest_end +%= 1;
    const slice: []const comptime_int = buf[0..dest_end];
    const ptr: [*]const comptime_int = slice.ptr;
    _ = ptr[0..5];
}
export fn start() void {
    comptime var buf: [1]comptime_int = .{0};
    var dest_start: usize = 0;
    dest_start +%= 1;
    const slice: []const comptime_int = buf[dest_start..][0..9];
    _ = slice[0];
}
export fn len() void {
    comptime var buf: [1]comptime_int = .{0};
    var dest_len: usize = 0;
    dest_len +%= 1;
    const slice: []const comptime_int = buf[0..][0..dest_len];
    _ = slice[0];
}
export fn end_const() void {
    const buf: [1]type = .{void};
    var dest_end: usize = 0;
    dest_end +%= 1;
    const slice: []const type = buf[0..dest_end];
    _ = slice[0];
}
export fn start_const() void {
    const buf: [1]type = .{void};
    var dest_start: usize = 0;
    dest_start +%= 1;
    const slice: []const type = buf[dest_start..][0..9];
    _ = slice[0];
}
export fn len_const() void {
    const buf: [1]type = .{void};
    var dest_len: usize = 0;
    dest_len +%= 1;
    const slice: []const type = buf[0..][0..dest_len];
    _ = slice[0];
}
```
#### Output (master)
```
zig build-obj runtime_indices_slicing_comptime_only_pointers.zig
Trace/breakpoint trap
```

This crash is produced by example function `end`. If `end` is rewritten to be like the other functions, the output is several apparently self-contradictory compile errors:

```
zig build-obj runtime_indices_slicing_comptime_only_pointers.zig
runtime_indices_slicing_comptime_only_pointers.zig:6:15: error: values of type '[]const comptime_int' must be comptime-known, but index value is runtime-known
    _ = slice[0];
              ^
runtime_indices_slicing_comptime_only_pointers.zig:13:15: error: values of type '[]const comptime_int' must be comptime-known, but index value is runtime-known
    _ = slice[0];
              ^
runtime_indices_slicing_comptime_only_pointers.zig:20:15: error: values of type '[]const comptime_int' must be comptime-known, but index value is runtime-known
    _ = slice[0];
              ^
runtime_indices_slicing_comptime_only_pointers.zig:27:15: error: values of type '[]const type' must be comptime-known, but index value is runtime-known
    _ = slice[0];
              ^
runtime_indices_slicing_comptime_only_pointers.zig:27:14: note: types are not available at runtime
    _ = slice[0];
        ~~~~~^~~
runtime_indices_slicing_comptime_only_pointers.zig:34:15: error: values of type '[]const type' must be comptime-known, but index value is runtime-known
    _ = slice[0];
              ^
runtime_indices_slicing_comptime_only_pointers.zig:34:14: note: types are not available at runtime
    _ = slice[0];
        ~~~~~^~~
runtime_indices_slicing_comptime_only_pointers.zig:41:15: error: values of type '[]const type' must be comptime-known, but index value is runtime-known
    _ = slice[0];
              ^
runtime_indices_slicing_comptime_only_pointers.zig:41:14: note: types are not available at runtime
    _ = slice[0];
        ~~~~~^~~
```
#### Output (slices_1)

This is what the new compile errors look like:

```
zig build-obj runtime_indices_slicing_comptime_only_pointers.zig
runtime_indices_slicing_comptime_only_pointers.zig:5:44: error: unable to resolve operand slicing comptime-only type '[1]comptime_int'
    const slice: []const comptime_int = buf[0..dest_end];
                                        ~~~^~~~~~~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:5:48: note: end index must be comptime-known
    const slice: []const comptime_int = buf[0..dest_end];
                                               ^~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:13:58: error: unable to resolve operand slicing comptime-only type '[1]comptime_int'
    const slice: []const comptime_int = buf[dest_start..][0..9];
                                        ~~~~~~~~~~~~~~~~~^~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:13:45: note: start index must be comptime-known
    const slice: []const comptime_int = buf[dest_start..][0..9];
                                            ^~~~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:20:49: error: unable to resolve operand slicing comptime-only type '[1]comptime_int'
    const slice: []const comptime_int = buf[0..][0..dest_len];
                                        ~~~~~~~~^~~~~~~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:20:53: note: length must be comptime-known
    const slice: []const comptime_int = buf[0..][0..dest_len];
                                                    ^~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:27:36: error: unable to resolve operand slicing comptime-only type '[1]type'
    const slice: []const type = buf[0..dest_end];
                                ~~~^~~~~~~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:27:40: note: end index must be comptime-known
    const slice: []const type = buf[0..dest_end];
                                       ^~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:34:50: error: unable to resolve operand slicing comptime-only type '[1]type'
    const slice: []const type = buf[dest_start..][0..9];
                                ~~~~~~~~~~~~~~~~~^~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:34:37: note: start index must be comptime-known
    const slice: []const type = buf[dest_start..][0..9];
                                    ^~~~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:41:41: error: unable to resolve operand slicing comptime-only type '[1]type'
    const slice: []const type = buf[0..][0..dest_len];
                                ~~~~~~~~^~~~~~~~~~~~~
runtime_indices_slicing_comptime_only_pointers.zig:41:45: note: length must be comptime-known
    const slice: []const type = buf[0..][0..dest_len];
                                            ^~~~~~~~
```